### PR TITLE
#2439 Do not throw NPE getting accessors for null typeElement

### LINKED ISSUE #2439: With wrong config like Enum[] to String， the mapping processor throw NPE
- issue
With wrong config like Enum[] to String， the mapping processor throw NPE.
Mapper code：
```
@Mapper
public interface LiveMapper {
    LiveMapper INSTANCE = Mappers.getMapper(LiveMapper.class);
    @Mappings({
             @Mapping(target = "statusName", source = "status.desc"), 
            @Mapping(target = "modeName", source = "mode.desc") 
    })
    LiveVo from(Live entity);
}

```

LiveVo code：
```
public class LiveVo {
    private LiveMode[] mode;  
    private String modeName;  
    private LiveStatus status;  
    private String statusName;  
}  
```

Live coe：
```
public class Live{
    private LiveMode[] mode;
    private LiveStatus status;
}
```

version：
    mspstruct：1.3.1.Final
    processor：1.2.0.Beta2
NPE info：
`java: Internal error in the mapping processor: 
java.lang.NullPointerException  	
at org.mapstruct.ap.internal.util.workarounds.SpecificCompilerWorkarounds.replaceTypeElementIfNecessary(SpecificCompilerWorkarounds.java:99)  	
at org.mapstruct.ap.internal.util.Executables.getAllEnclosedExecutableElements(Executables.java:99)  	
at org.mapstruct.ap.internal.model.common.Type.getAllMethods(Type.java:633)  	
at org.mapstruct.ap.internal.model.common.Type.getPropertyReadAccessors(Type.java:496)  	
at org.mapstruct.ap.internal.model.source.SourceReference$BuilderFromMapping.getSourceEntries(SourceReference.java:210)  	
at org.mapstruct.ap.internal.model.source.SourceReference$BuilderFromMapping.build(SourceReference.java:155)  	
at org.mapstruct.ap.internal.model.source.Mapping.init(Mapping.java:436)  	
at org.mapstruct.ap.internal.model.source.Mapping.init(Mapping.java:413)  	
at org.mapstruct.ap.internal.model.source.SourceMethod$Builder.build(SourceMethod.java:200)  	
at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.getMethodRequiringImplementation(MethodRetrievalProcessor.java:261)  	
at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.getMethod(MethodRetrievalProcessor.java:188)  	
at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.retrieveMethods(MethodRetrievalProcessor.java:143)  	
at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.process(MethodRetrievalProcessor.java:85)  	
at org.mapstruct.ap.internal.processor.MethodRetrievalProcessor.process(MethodRetrievalProcessor.java:59)  	
at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:283)  	
at org.mapstruct.ap.MappingProcessor.processMapperTypeElement(MappingProcessor.java:263)  	
at org.mapstruct.ap.MappingProcessor.processMapperElements(MappingProcessor.java:221)  	
at org.mapstruct.ap.MappingProcessor.process(MappingProcessor.java:156)  	
at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)  	
at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:705)  	
at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)  	
at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1035)  	
at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1176)  	
at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)  	
at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)  	
at com.sun.tools.javac.main.Main.compile(Main.java:523)  	
at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)  	
at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)  	
at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:207)  	
at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:491)  	
at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:343)  	
at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:268)  	
at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:221)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1448)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:1126)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1193)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:1087)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:853)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:443)  	
at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:184)  	
at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:135)  	
at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:294)  	
at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:127)  	
at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:218)  	
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)  	
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)  	
at java.lang.Thread.run(Thread.java:748)  `

### PR DESCRIPTION
**Pull request by filiphr**
Provide better error message if most similar word is null, i.e. type has no properties

Fixes #2439

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 1]
```diff
@@ -70,6 +70,7 @@
     PROPERTYMAPPING_INVALID_PARAMETER_NAME( "Method has no source parameter named \"%s\". Method source parameters are: \"%s\"." ),
     PROPERTYMAPPING_NO_PROPERTY_IN_PARAMETER( "The type of parameter \"%s\" has no property named \"%s\"." ),
     PROPERTYMAPPING_INVALID_PROPERTY_NAME( "No property named \"%s\" exists in source parameter(s). Did you mean \"%s\"?" ),
+    PROPERTYMAPPING_INVALID_PROPERTY_NAME_SOURCE_HAS_NO_PROPERTIES( "No property named \"%s\" exists in source parameter(s). Type \"%s\" has no properties." ),
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:73]: `has no properties`? That's not certain, is it?

**Review by sjaakd — COMMENTED** [review]
bit uncertain on error message

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 1 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:73]: We are certain. `mostSimilarWord` can be `null` only if there were no other `similarWords`

https://github.com/mapstruct/mapstruct/blob/857f87276f686f0659b0fa604aa363e25701916b/processor/src/main/java/org/mapstruct/ap/internal/util/Strings.java#L217-L228

No other `similarWords` can happen if `sourceType.getPropertyReadAccessors().keySet()` is empty. If that is empty then there are no read properties. Perhaps we need to add that?

File: processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/SourceReference.java [hunk 2]
```diff
@@ -281,13 +281,22 @@ private void reportErrorOnNoMatch( Parameter parameter, String[] propertyNames,
...
                     sourceType.getPropertyReadAccessors().keySet()
                 );
-                List<String> elements = new ArrayList<>(
-                    Arrays.asList( propertyNames ).subList( 0, notFoundPropertyIndex )
-                );
-                elements.add( mostSimilarWord );
-                reportMappingError(
-                    Message.PROPERTYMAPPING_INVALID_PROPERTY_NAME, sourceName, Strings.join( elements, "." )
-                );
+                if ( mostSimilarWord != null ) {
```
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/SourceReference.java:284]: I guess that  reeds a lot better when you do it like this:
```java
if ( !sourceType.getPropertyReadAccessors().keySet().isEmpty() ) {

    String mostSimilarWord = Strings.getMostSimilarWord(...  ) 
    ... 
}
else { 
  ...
} 
```

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java (hunk 1 above)
Comment (sjaakd) [reply processor/src/main/java/org/mapstruct/ap/internal/util/Message.java:73]: Ah.. I understand you now.. but check my comment above..

**Review by sjaakd — CHANGES_REQUESTED** [review]
see my single comments

File: processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/SourceReference.java (hunk 2 above)
Comment (filiphr) [reply processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/SourceReference.java:284]: You are right. I'll change it to this

**Pull request merged**